### PR TITLE
refractor(user module) password's strength validation and unique username

### PR DESCRIPTION
**Pull request by jeremielodi**
This PR provides a password's strength validation directive and notify the user if the current user name he has typed to create a new user is already used because it should be unique

### CODE REVIEW

File: client/src/js/components/bhStrengthMeter.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+angular.module('bhima.components')
+.component('bhStrengthMeter', {
+  template :
+    "<div class='strength-meter'>"+
+    " <div class='strength-meter-fill' data-strength='{{$ctrl.counter()}}'></div>"+
+    "</div>",
+  bindings : {
+    password : '=',
```
Comment (jniles) [inline client/src/js/components/bhStrengthMeter.js]: Is there any reason for this to be a [two-way binding](https://docs.angularjs.org/guide/component#component-based-application-architecture)?  Can it be a one-way binding (`<`)?

File: client/src/js/directives/bhPasswordMeter.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+PasswordValidatorDirective.$inject = ['PasswordMeterService'];
+
+/**
+ * Passord Validator Directive
+ *
+ * This directive provides a generic password validatation.
+ * the validation condition is defined at  PasswordMeterService
+ *
+ * @example
```
Comment (jniles) [inline client/src/js/directives/bhPasswordMeter.js:12]: Nice comments!

File: client/src/js/services/PasswordMeterService.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
+angular.module('bhima.services')
+.service('PasswordMeterService', PasswordMeterService);
+
+PasswordMeterService.$inject = [];
```
Comment (jniles) [inline client/src/js/services/PasswordMeterService.js]: If you don't inject anything, there is no need to put this empty array up here.

File: client/src/modules/users/user.modal.html [hunk 4]
```diff
@@ -12,16 +12,28 @@
...
+  <div class="modal-body" style="overflow-y: scroll; max-height:600px; ">
+
 
     <div class="form-group" ng-class="{ 'has-error' : UserForm.$submitted && UserForm.display_name.$invalid }">
-      <label class="control-label" translate>FORM.LABELS.USERNAME</label>
-      <input name="display_name" ng-model="UserModalCtrl.user.display_name" autocomplete="off" ng-maxlength="UserModalCtrl.maxLength" class="form-control" required>
-      <div class="help-block" ng-messages="UserForm.display_name.$error" ng-show="UserForm.$submitted">
-        <div ng-messages-include="modules/templates/messages.tmpl.html"></div>
+      <label class="control-label col-md-3" translate>FORM.LABELS.USERNAME</label>
+      <div class="col-md-12">
```
Comment (jniles) [inline client/src/modules/users/user.modal.html]: Having a `col-md-3` and `col-md-12` on the same line doesn't make any sense ... bootstrap only has 12 columns.  Which do you mean to put here - `col-md-3` and `col-md-9`?

File: client/src/modules/users/user.modal.html [hunk 5]
```diff
@@ -12,16 +12,28 @@
...
+
+
+        <div class="help-block" ng-messages="UserForm.display_name.$error" ng-show="UserForm.$submitted">
+          <div ng-messages-include="modules/templates/messages.tmpl.html"></div>
+        </div>
       </div>
     </div>
 
+    <ng-messages for="UserForm.display_name.$error" ng-if="UserForm.display_name.$dirty">
+      <ng-message when="uniqueUserName"> user already exists</ng-message>
```
Comment (jniles) [inline client/src/modules/users/user.modal.html]: This should be translated.

File: client/src/modules/users/user.modal.html [hunk 6]
```diff
@@ -73,14 +85,21 @@
...
+        <input
+          name="password"
+          ng-model="UserModalCtrl.user.password"
+          class="form-control"
+          bh-password-meter
+          type="password" required>
 
-        <span ng-show="UserForm.password.$dirty && !UserModalCtrl.validPassword()" class="glyphicon glyphicon-warning-sign form-control-feedback text-danger" aria-hidden="true"></span>
-        <span ng-show="UserForm.password.$dirty && UserModalCtrl.validPassword()" class="glyphicon glyphicon-ok form-control-feedback text-success" aria-hidden="true"></span>
+          <bh-strength-meter password="UserForm.password.$viewValue"></bh-strength-meter>
```
Comment (jniles) [inline client/src/modules/users/user.modal.html:93]: :+1:

File: client/src/modules/users/user.modal.js [hunk 7]
```diff
@@ -11,21 +14,22 @@ function UserModalController($state, Projects, Users, Notify, AppCache) {
   // the user object that is either edited or created
   vm.user = {};
   vm.stateParams = {};
+  vm.passwordValidationStrength = '';
```
Comment (jniles) [inline client/src/modules/users/user.modal.js]: You never use this - it can be removed.

File: client/src/modules/users/user.modal.js [hunk 8]
```diff
@@ -11,21 +14,22 @@ function UserModalController($state, Projects, Users, Notify, AppCache) {
...
-
-
   vm.isCreating = vm.stateParams.creating;
 
-  //exposed methods
-  vm.submit = submit;
-  vm.closeModal = closeModal;
-  vm.validPassword = validPassword;
-  vm.editPassword = editPassword;
+  vm.passwordValidationCount = -1;
```
Comment (jniles) [inline client/src/modules/users/user.modal.js]: It doesn't look like you use this either.

File: server/config/routes.js [hunk 9]
```diff
@@ -438,6 +438,7 @@ exports.configure = function configure(app) {
   app.get('/users', users.list);
   app.post('/users', users.create);
   app.get('/users/:id', users.detail);
+  app.get('/users/:username/exists', users.exists);
```
Comment (jniles) [inline server/config/routes.js:441]: :+1:

File: server/controllers/admin/users/index.js [hunk 10]
```diff
@@ -124,6 +125,17 @@ function detail(req, res, next) {
   .done();
 }
 
+function exists(req, res, next) {
+  const sql='SELECT count(id) as nbr FROM user WHERE username=?';
```
Comment (jniles) [inline server/controllers/admin/users/index.js:129]: A more efficient SQL query might be `SELECT true FROM user WHERE username = ?`.

**Review by jniles — COMMENTED** [review]
This is really cool!  There are a few clean up things to take care of.

File: client/src/modules/users/user.modal.html (hunk 4 above)
Comment (jeremielodi) [reply client/src/modules/users/user.modal.html]: the curent disposition of the form is : the input is underneath the lable that is why it should have this class 'col-md-12', I'm change the label class col-md-3 to col-md-12 too.
 The 'col-md-12' class in the username input help to set the seach icon visible

Comment (jniles) [reply client/src/modules/users/user.modal.html]: It doesn't work without the `col-md-12`?  That's interesting.

Comment (jeremielodi) [reply client/src/modules/users/user.modal.html]: It works but the seach icon position is bad

**Comment by jniles** [comment]
LGTM.  bors r+

**Pull request merged**